File: leetcode/solutions/0238.ProductOfArrayExceptSelf/productExceptSelf.ts
export default function productExceptSelf(nums: number[]): number[] {
  let left = 1;
  let right = 1;
  const size = nums.length;
  const result = Array.from({ length: size }, () => 1);

  for (let l = 0; l < size; l++) {
    const r = size - l - 1;
    result[l] *= left;
    result[r] *= right;
    left *= nums[l];
    right *= nums[r];
  }

  return result;
}

File: leetcode/solutions/0238.ProductOfArrayExceptSelf/productExceptSelf.go
package solution

func productExceptSelf(nums []int) []int {
	left, right := 1, 1
	res := make([]int, len(nums))
	for i := range res {
		res[i] = 1
	}

	for l := range nums {
		r := len(nums) - l - 1
		res[l] *= left
		res[r] *= right
		left, right = left*nums[l], right*nums[r]
	}

	return res
}
